Add text similarity reranker retriever

demjened opened the pull request:
This PR adds support for the `text_similarity_reranker` retriever in the `inference` XPack plugin. It builds upon the recently added [RankFeature search phase](https://github.com/elastic/elasticsearch/pull/107099) and related framework that allows us to define querying and ranking logic at the shard and coordinator level.

Key changes:
* `TextSimilarityRankRetrieverBuilder` is a `RetrieverBuilder` that defines the fields and the parser for text similarity reranking: `field` (which field to extract from the first stage docs for reranking), `inference_id` and `inference_text` (inference configuration), `rank_window_size`, and an optional `min_score`.
* `TextSimilarityRankBuilder` is a `RankBuilder` that accepts the above parameters and creates a specific `TextSimilarityRankFeaturePhaseRankCoordinatorContext`.
* `TextSimilarityRankFeaturePhaseRankCoordinatorContext` defines the reranking logic for a coordinator node: extract feature fields from fetched docs, send them to a rerank inference endpoint, and filter and reorder the documents according to the returned scores.

Example request (assuming the `my-cohere-rerank-model-v3` inference endpoint has been set up):
```json
POST _search
{
  "retriever": {
    "text_similarity_reranker": {
      "retriever": {
        "standard": {
          "query": {
            "match": {
              "text": "How often does the moon hide the sun?"
            }
          }
        }
      },
      "field": "text",
      "inference_id": "my-cohere-rerank-model-v3",
      "inference_text": "How often does the moon hide the sun?",
      "rank_window_size": 100,
      "min_score": 0.5
    }
  },
  "_source": ["title", "text"],
  "size": 10
}
```

Example response:
```
{
  "took": 191,
  ...
  "hits": {
    "total": {
      "value": 10000,
      "relation": "gte"
    },
    "max_score": 0.84312165,
    "hits": [
      {
        "_index": "wikipedia_elser_demo",
        "_id": "17800",
        "_score": 0.84312165,
        "_rank": 1,
        "_ignored": [
          "text.keyword"
        ],
        "_source": {
          "title": "Solar eclipse",
          "text": "As seen from Earth, a solar eclipse /\"ee-klips\"/ happens when the Moon is directly between the Earth and the Sun. This makes the Moon fully or partially (partly) cover the sun. Solar eclipses can only happen during a new moon. Every year there are about two solar eclipses. Sometimes there are even five solar eclipses in a year. However, only two of these can be total solar eclipses, and often a year will pass without a total eclipse."
        }
      },
     ...
    ]
  }
}
```

The explain hit functionality is also implemented here.

demjened committed:
Add text similarity rerank builder and context

demjened committed:
Add min_score support

demjened committed:
Add tests

demjened committed:
Filter by min_score + rename rank_window_size

demjened committed:
Add transport version

demjened commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankTests.java:
```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.xpack.core.inference.action.InferenceAction;
+import org.elasticsearch.xpack.core.inference.results.RankedDocsResults;
+import org.elasticsearch.xpack.inference.InferencePlugin;
```
This import cannot be found, I guess it's not declared in the module definition. IntelliJ offers to fix it so it compiles locally, but it doesn't change anything in the code, and the CI still fails. Any thoughts on how I can import the module that contains `InferencePlugin`?

demjened replied:
As per the [Slack discussion](https://elastic.slack.com/archives/C0D8ST60Y/p1718742644076469) I'll move the test class into the inference module to resolve this.

demjened commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankFeaturePhaseRankCoordinatorContextTest.java:
```diff
@@ -0,0 +1,47 @@
+
+import org.elasticsearch.test.ESTestCase;
+
+public class TextSimilarityRankFeaturePhaseRankCoordinatorContextTest extends ESTestCase {
```
WIP, working on these tests.

demjened committed:
Add doc + unit test

pmpailis commented on server/src/internalClusterTest/java/org/elasticsearch/search/rank/MockedRequestActionBasedRerankerIT.java:
```diff
@@ -321,33 +333,38 @@ public MockRequestActionBasedRankBuilder(
             out.writeString(field);
             out.writeString(inferenceId);
             out.writeString(inferenceText);
+            out.writeFloat(minScore);
```
I believe that usually, we'd have to conditionally write new fields depending on the `TransportVersion` but since this completely new functionality (not used even in serverless) we should be ok.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    public String getWriteableName() {
+        return "text_similarity_rank_builder";
```
NIT: could export this to a constant

demjened replied:
Done

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    public void doXContent(XContentBuilder builder, Params params) throws IOException {
+        throw new UnsupportedOperationException("not supported");
```
Is this a `TODO` or don't we need to implement this?

demjened replied:
Added implementation.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    public Explanation explainHit(Explanation baseExplanation, RankDoc scoreDoc, List<String> queryNames) {
+        return null;
```
We could implement something similar to rrf's explain to provide explain info on the match (e..g include info on the inferenceId, initial score etc).

demjened replied:
Sounds good, I'm working on this.

demjened replied:
`GET /wikipedia_elser_demo/_explain/17800` gives me a parsing exception: `request does not support [retriever]`. Is there any other place we need to declare the supported fields for the `_explain` API?

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    protected boolean doEquals(RankBuilder other) {
+        return false;
```
`todo` :)

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    protected int doHashCode() {
+        return 0;
```
`todo` :)

demjened replied:
Added implementation.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -0,0 +1,97 @@
+    protected final Client client;
+    protected final String inferenceId;
+    protected final String inferenceText;
+    protected final float minScore;
```
Maybe we could move `minScore` to `RankFeaturePhaseRankCoordinatorContext` ?

demjened replied:
Not sure about this, this would assume that all reranking implementation operate with scores. I prefer to leave this for a future implementation, if we see some common logic. Let me know if you have a strong opinion about this.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,125 @@
+        this.inferenceId = in.readString();
+        this.inferenceText = in.readString();
+        this.field = in.readString();
+        this.minScore = in.readFloat();
```
IIUC if we don't specify a `minScore` we would have a neg inf value, right? Maybe we could use `Float` instead and just have it as null and read/write `optionalFloat` instead? (+ maybe we could also return true for the predicate in this scenario w/o any comparisons)?

demjened replied:
Switched to optional Float.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,132 @@
+    protected void doToXContent(XContentBuilder builder, Params params) {}
+
+    @Override
+    protected boolean doEquals(Object o) {
```
`todo`

demjened replied:
Added implementation.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    protected void doToXContent(XContentBuilder builder, Params params) {}
```
`todo`

demjened replied:
Added implementation.

pmpailis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/rerank/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    protected int doHashCode() {
+        return 0;
```
`todo`

demjened replied:
Added implementation.

pmpailis reviewed (commented):
Nice work @demjened ! 🚀   Added some minor comments and some `todo` just to keep track :)

demjened committed:
Implement equals and hashcode, use constant

demjened committed:
Use optional float for min_score

demjened committed:
Implement explainHit

demjened committed:
Extract TS reranking to its own plugin

demjened committed:
Fix description

demjened committed:
Move field name

demjened commented on x-pack/plugin/rank-text-similarity/build.gradle:
```diff
@@ -0,0 +1,31 @@
+}
+
+dependencies {
+//  compileOnly project(path: xpackModule('core'))
```
I'm not familiar with this DSL, so I probably made some mistakes here.
- What is the difference between `project('...')` and `project(path: xpackModule(...))`?
- When should I use `testArtifact(project())` vs just `project()` under `testImplementation`?
- What is `clusterModules` for?

rjernst replied:
You need to use the `project(path:...)` variant. That comes from the `DependencyHandler` in gradle. The other one grabs a project object from Gradle, but that's a different thing.

`testArtifact` is for when depending on the test classes from another project. That other project has to be setup specially to export its test artifact.

`clusterModules` is for the "new" test clusters framework. This ties another project as a module that will be built, for use by java rest tests. It's only available when you have applied the java rest test plugin, ie `apply plugin: 'elasticsearch.internal-java-rest-test'` at the top of `build.gradle`.

demjened committed:
Add featurespec, fix build

demjened committed:
Fix dir

demjened committed:
Add XContent parsing, clean up build.gradle

demjened committed:
Add license check

demjened committed:
Spotless

demjened committed:
Remove clusterModules from gradle config

demjened commented on x-pack/plugin/rank-text-similarity/build.gradle:
```diff
@@ -0,0 +1,27 @@
+  testImplementation(testArtifact(project(xpackModule('core'))))
+  testImplementation project(path: xpackModule('inference'))
+
+  clusterModules project(xpackModule('rank-text-similarity'))
```
Thanks for calling it out - I removed it.

demjened committed:
Fix multi-node tests

demjened commented on x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle:
```diff
@@ -15,6 +15,7 @@ dependencies {
   javaRestTestImplementation project(path: xpackModule('monitoring'))
   javaRestTestImplementation project(path: xpackModule('transform'))
   javaRestTestImplementation project(path: xpackModule('rank-rrf'))
+  javaRestTestImplementation project(path: xpackModule('rank-text-similarity'))
```
CI was [failing](https://buildkite.com/elastic/elasticsearch-pull-request/builds/23426#01903b1b-a5fa-48f7-acd2-b1b7e0a4fac4) - found a possible resolution similar to the one for `rank-rrf` as per [Slack thread](https://elastic.slack.com/archives/C8UUBNASY/p1710426954194729?thread_ts=1710316157.288789&cid=C8UUBNASY).

demjened commented:
@mark-vieira I addressed your comments, could you re-review please?

benwtrent commented on x-pack/plugin/rank-text-similarity/src/main/java/org/elasticsearch/xpack/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,142 @@
+
+        PARSER.declareField((r, v) -> r.inferenceText = v, (p, c) -> {
+            String inferenceText = p.text();
+            c.trackSectionUsage(NAME + ":" + INFERENCE_TEXT_FIELD.getPreferredName());
```
Why are we tracking every individual section?

demjened replied:
Again, I was following a pattern I've seen in `StandardRetrieverBuilder`, where [every section is being tracked](https://github.com/elastic/elasticsearch/blob/b0d9f95e534e07aca216a556bdaa5e137fe87bbd/server/src/main/java/org/elasticsearch/search/retriever/StandardRetrieverBuilder.java#L56).

Telemetry requirements are going to be refined after 8.15. Until then I can remove the tracking of those fields that are mandatory in this retriever as it's redundant.

demjened replied:
Removed tracking of sections.

demjened commented on x-pack/plugin/rank-text-similarity/src/main/java/org/elasticsearch/xpack/rank/textsimilarity/TextSimilarityRankPlugin.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.List;
+
+public class TextSimilarityRankPlugin extends Plugin implements SearchPlugin {
```
I mirrored the RRF plugin, I wasn't aware of any strong arguments for not having separate plugins.
Is it OK if I file a tech debt ticket to move these into `inference` along with RRF, rather than addressing it here?

demjened committed:
Simplify

demjened commented on x-pack/plugin/rank-text-similarity/src/main/java/org/elasticsearch/xpack/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -0,0 +1,97 @@
+        InferenceAction.Request request = generateRequest(featureData);
+        try {
+            ActionType<InferenceAction.Response> action = InferenceAction.INSTANCE;
+            client.execute(action, request, actionListener);
```
Done.

demjened committed:
Move feature into inference plugin

demjened committed:
Add Javadoc

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferencePlugin.java:
```diff
@@ -208,11 +214,22 @@ public List<InferenceServiceExtension.Factory> getInferenceServiceFactories() {
     public List<NamedWriteableRegistry.Entry> getNamedWriteables() {
-        var entries = new ArrayList<NamedWriteableRegistry.Entry>();
-        entries.addAll(InferenceNamedWriteablesProvider.getNamedWriteables());
+        var entries = new ArrayList<>(InferenceNamedWriteablesProvider.getNamedWriteables());
+        entries.add(new NamedWriteableRegistry.Entry(RankBuilder.class, TextSimilarityRankBuilder.NAME, TextSimilarityRankBuilder::new));
         return entries;
     }
 
+    @Override
+    public List<NamedXContentRegistry.Entry> getNamedXContent() {
+        return List.of(
+            new NamedXContentRegistry.Entry(
+                RankBuilder.class,
+                new ParseField(TextSimilarityRankBuilder.NAME),
+                TextSimilarityRankBuilder::fromXContent
+            )
+        );
+    }
+
```
I am confused on the difference between a "RankBuilder" and a "RetrieverBuilder"

The retriever spec registration adds named xcontent automatically via
```
    private void registerRetriever(RetrieverSpec<?> spec) {
```

@pmpailis does that not work anylonger?

pmpailis replied:
For parsing, I think that just the export through the parent `TextSimilarityRankRetrieverBuilder` [here](https://github.com/elastic/elasticsearch/blob/f1155a554d885e010092225b86887947c4e30c76/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferencePlugin.java#L329), should be sufficient enough, as we don't expect to parse directly any XContent for the  `RankBuilder`. We do need however the `NamedWriteable`  as we pass the `RankBuilder` through `source`. 

Also, just for completeness, a `RetrieverBuilder` is the top-level element in the tree and is one of `knn`, `standard`, `rrf`, and now, `text-similarity-rank`. The `RankBuilder` on the other hand is something specific that only `rrf` and `text-similarity-rank` retrievers specify and export to source, and defines the necessary steps to apply any custom ranking. So eventually, as the retriever will be flatten out to be added to source, the `rankBuilder` will be left as a primary object of `source` that needs to be (de)serialized.  

It's quite confusing and the concepts are tied together in a not-so-clear way, but hopefully this will be much more clear with the pending rework and clear distinction of the `retriever` and `ranker` concepts.

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,138 @@
+    static {
+        PARSER.declareObject((r, v) -> r.retrieverBuilder = v, (p, c) -> {
+            RetrieverBuilder retrieverBuilder = RetrieverBuilder.parseInnerRetrieverBuilder(p, c);
+            c.trackSectionUsage(TextSimilarityRankBuilder.NAME + ":" + RETRIEVER_FIELD.getPreferredName());
```
We shouldn't do this track usage for all these fields individually. Its gonna add so much noise. 

Right now we only care about the similarity ranker being used, we don't care if folks set window size, etc. This is really overkill.

demjened replied:
Removed all tracking from here. The retriever builder already tracks the usage on specific retrievers, e.g. `standard` or `text_similarity_reranker`.

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankTests.java:
```diff
@@ -0,0 +1,154 @@
+
+import static java.util.Collections.singletonList;
+
+public class TextSimilarityRankTests extends ESSingleNodeTestCase {
```
could we have a test where inference fails?

demjened replied:
Added a test case: https://github.com/elastic/elasticsearch/pull/109813/commits/6c85d68ba3461dba615692010bcd8e33c59d7907.

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -0,0 +1,99 @@
+        // and then pass the results
+        final ActionListener<InferenceAction.Response> actionListener = scoreListener.delegateFailureAndWrap((l, r) -> {
+            float[] scores = extractScoresFromResponse(r);
+            assert scores.length == featureDocs.length;
```
this should be a hard failure. Otherwise, we have no way of knowing which score corresponds to which doc.

demjened replied:
Updated logic to fail with exception.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/context/RankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -72,13 +81,18 @@ public void computeRankScoresForGlobalResults(
     public RankFeatureDoc[] rankAndPaginate(RankFeatureDoc[] rankFeatureDocs) {
-        Arrays.sort(rankFeatureDocs, Comparator.comparing((RankFeatureDoc doc) -> doc.score).reversed());
-        RankFeatureDoc[] topResults = new RankFeatureDoc[Math.max(0, Math.min(size, rankFeatureDocs.length - from))];
+        RankFeatureDoc[] rankFeatureDocsAboveMinScore = Arrays.stream(rankFeatureDocs)
+            .filter(this::keepRankFeatureDoc)
```
this is misnamed, its about if we keep the feature doc or not. There is nothing in particular about "min score" here. All the concern about min score seems to be in the sub class.

demjened replied:
[Refactored](https://github.com/elastic/elasticsearch/pull/109813/commits/62ab07cb299af3ecf551fa2021c2542c940b9271) - I prefer to move `min_score`-based logic into the subclass, since it's not guaranteed that every `RankFeaturePhaseRankCoordinatorContext` implementation will use scores.

benwtrent reviewed (commented):
Is there anyway to get yaml tests for this? I am not sure if the inference plugin has any test infrastructure for this or not. 

I am assuming docs will be in a forth coming PR.

demjened committed:
Update docs/changelog/109813.yaml

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @demjened, I've created a changelog YAML for you.

demjened committed:
Remove tracking of unnecessary sections

demjened committed:
Hard failure on doc count != score count

demjened committed:
Remove fromXContent parsing

demjened committed:
Fix leak in tests

demjened committed:
Add failure test

demjened committed:
Remove minscore tracking

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankTests.java:
```diff
@@ -0,0 +1,180 @@
+        );
+    }
+
+    public void testRerankInferenceFailure() {
```
Thank you! Could we have a test now that ensures we throw when we don't get the right number of rankings from the inference service?

demjened replied:
Added a test case (https://github.com/elastic/elasticsearch/pull/109813/commits/a8564a2e0cdf94e5b19909fbe8479aa707a4dead).
I feel the in-place behavior change for the static `TestFilter` is a bit clunky, but I don't have a better idea as it's initialized once.

pmpailis replied:
I think that there is no guarantee of test order execution, so maybe we could be met with a scenario where this runs first, changes the `runMode`, and then the first couple of tests  (e.g. `testRerankWithMinScore`) which assume `TestFilterRunMode.NORMAL`, might fail as a result of this?

pmpailis replied:
Another option could be to override the `TextSimilarityRankFeaturePhaseRankCoordinatorContext#generateRequest` and pass any needed metadata/flags along with the `taskSettings` configuration.

demjened replied:
I did some refactoring: extracted `TextSimilarityTestPlugin` into its own class and added support for a context that returns invalid inference results for that test (https://github.com/elastic/elasticsearch/pull/109813/commits/444bf09c0388698c69abc1b79f303eeecb8d0bf5).
Now `TextSimilarityRankTests` and `TextSimilarityRankMultiNodeTests` both use this plugin.

demjened committed:
Put back node feature

demjened committed:
Switch from platinum to enterprise license

demjened committed:
Encapsulate min_score filtering in subclass

demjened commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferenceFeatures.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public Set<NodeFeature> getFeatures() {
+        return Set.of(TextSimilarityRankRetrieverBuilder.TEXT_SIMILARITY_RERANKER_RETRIEVER_SUPPORTED);
```
@benwtrent @pmpailis I added this feature spec to the `inference` plugin so that we can [check the feature and fail fast in a mixed-cluster scenario](https://github.com/elastic/elasticsearch/pull/109813/files#diff-d1390fdfd7cd709bf892434e8fe8192cd055e8b55efd77174fa580d88e2542a8R66-R68).

demjened committed:
Add test case for inference response mismatch

demjened committed:
Spotless

pmpailis commented on server/src/main/java/org/elasticsearch/search/rank/context/RankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -44,6 +45,16 @@ public RankFeaturePhaseRankCoordinatorContext(int size, int from, int rankWindow
+     * Filters and sorts the provided documents.
+     * @param originalDocs documents to process
+     */
+    protected RankFeatureDoc[] filterAndSort(RankFeatureDoc[] originalDocs) {
```
Minor, but maybe we could rename this as in this base class there's no filtering taking place.

demjened replied:
Renamed to `preprocess()` and added Javadoc (https://github.com/elastic/elasticsearch/pull/109813/commits/3b18dc0fe7f19d1614f7a6ad14cfe6b1948500f1).

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+    );
+
+    static {
+        PARSER.declareObject((r, v) -> r.retrieverBuilder = v, RetrieverBuilder::parseInnerRetrieverBuilder, RETRIEVER_FIELD);
+        PARSER.declareField((r, v) -> r.field = v, (p, c) -> p.text(), FIELD_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField(
+            (r, v) -> r.rankWindowSize = v,
+            (p, c) -> p.intValue(),
+            RANK_WINDOW_SIZE_FIELD,
+            ObjectParser.ValueType.INT_OR_NULL
+        );
+        PARSER.declareField((r, v) -> r.inferenceId = v, (p, c) -> p.text(), INFERENCE_ID_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.inferenceText = v, (p, c) -> p.text(), INFERENCE_TEXT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.minScore = v, (p, c) -> p.floatValue(), MIN_SCORE_FIELD, ObjectParser.ValueType.FLOAT_OR_NULL);
```
The way you have this parsing indicates all of these things are optional. This should not be allowed.

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+
+    static {
+        PARSER.declareObject((r, v) -> r.retrieverBuilder = v, RetrieverBuilder::parseInnerRetrieverBuilder, RETRIEVER_FIELD);
+        PARSER.declareField((r, v) -> r.field = v, (p, c) -> p.text(), FIELD_FIELD, ObjectParser.ValueType.STRING);
```
Please use `declareString`

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+            (r, v) -> r.rankWindowSize = v,
+            (p, c) -> p.intValue(),
+            RANK_WINDOW_SIZE_FIELD,
+            ObjectParser.ValueType.INT_OR_NULL
```
We shouldn't accept null, please use `declareInt`

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Override
+    protected void doToXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(RETRIEVER_FIELD.getPreferredName(), retrieverBuilder);
+        builder.field(FIELD_FIELD.getPreferredName(), field);
+        builder.field(RANK_WINDOW_SIZE_FIELD.getPreferredName(), rankWindowSize);
+        builder.field(MIN_SCORE_FIELD.getPreferredName(), minScore);
```
We should check for `null` where appropriate and not write out null values.

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankBuilder.java:
```diff
@@ -0,0 +1,192 @@
+        return new TextSimilarityRankBuilder(field, inferenceId, inferenceText, DEFAULT_RANK_WINDOW_SIZE, minScore);
+    });
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), INFERENCE_ID_FIELD);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), INFERENCE_TEXT_FIELD);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), FIELD_FIELD);
+        PARSER.declareFloat(ConstructingObjectParser.optionalConstructorArg(), MIN_SCORE_FIELD);
+    }
```
This parsing looks good :). Much better than the other object.

benwtrent reviewed (commented):
I thought this goes without saying, but, we need tests for serialization and parsing for any class that is serialized over the wire and parsed. 

Additionally, there should be tests that exercise the named writeable/xcontent parsing and serialization. I just noticed that there are none that do that.

So, there need to be `TextSimilarityRankRetrieverBuilderTests` & `TextSimilarityRankBuilderTests` both extending `AbstractSerializationTestCase`

I am also making an assumption that in following PRs we will:

 - Add performance profiling (indicating how long its taking for inference)
 - Add docs
 - Add multi-node integration tests (this is sorely needed)

demjened committed:
Add serialization test

demjened committed:
Switch retriever builder to ConstructingObjectParser

demjened commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+        );
+        PARSER.declareField((r, v) -> r.inferenceId = v, (p, c) -> p.text(), INFERENCE_ID_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.inferenceText = v, (p, c) -> p.text(), INFERENCE_TEXT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.minScore = v, (p, c) -> p.floatValue(), MIN_SCORE_FIELD, ObjectParser.ValueType.FLOAT_OR_NULL);
```
@benwtrent I switched the parser to use `ConstructingObjectParser` (https://github.com/elastic/elasticsearch/pull/109813/commits/199822e81485859876073e18a9e8631c728a38fc). 

The reason I was using the object parser was to enable tracking of sections (following [StandardRetrieverBuilder](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/retriever/StandardRetrieverBuilder.java#L54) as an example). But since that got removed, I'm switching back.

Hopefully this addresses your comments regards to parsing.

demjened committed:
Add retriever serialization test

demjened committed:
Add yaml test

demjened committed:
Exception if nesting other rank builder

demjened commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,123 @@
+    public void extractToSearchSourceBuilder(SearchSourceBuilder searchSourceBuilder, boolean compoundUsed) {
+        retrieverBuilder.extractToSearchSourceBuilder(searchSourceBuilder, compoundUsed);
+
+        if (searchSourceBuilder.rankBuilder() == null) {
```
@pmpailis Added code to throw exception in this case: https://github.com/elastic/elasticsearch/pull/109813/commits/cddc7cfe2fb7e4de002b39f9f9c73922f6f60f35.

Note it's not possible to reach this state because retriever nesting is limited to 2 levels deep, but it's good to have this safety net.

demjened commented:
@benwtrent Addressed some of your feedback, please review when you have a chance:
- Switched to `ConstructingObjectParser` that simplifies parser configuration
- Added serialization tests and yaml test (thanks @pmpailis!)
- Added code to throw if combining text similarity rank builder with other rank builder, as it's not supported

Still working on:
- Multi-node tests
- Docs (separate PR)

> Add performance profiling (indicating how long its taking for inference)

Do you mean benchmarking, or adding some logic in the code?

demjened committed:
Fix float rounding error

benwtrent commented:
> Add performance profiling (indicating how long its taking for inference)

@demjened we don't have this yet in retrievers 100%, but we need to track things when `profiling: true`. This will be a future PR.

demjened committed:
Fix multi-node tests

demjened committed:
Set trial license for yaml tests

demjened committed:
Spotless

demjened committed:
Fix trial license setting for yaml tests

demjened committed:
Update yaml test texts

benwtrent commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -0,0 +1,151 @@
+            String inferenceText = (String) args[2];
+            String field = (String) args[3];
+            int rankWindowSize = args[4] == null ? DEFAULT_RANK_WINDOW_SIZE : (int) args[4];
+            Float minScore = args[5] == null ? null : (float) args[5];
```
```suggestion
            Float minScore = (Float) args[5];
```

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankBuilderTests.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    protected TextSimilarityRankBuilder createTestInstance() {
+        return new TextSimilarityRankBuilder("my-field", "my-inference-id", "my-inference-text", randomIntBetween(1, 1000), randomFloat());
```
```suggestion
        return new TextSimilarityRankBuilder("my-field", "my-inference-id", "my-inference-text", randomIntBetween(1, 1000), randomBoolean() ? null : randomFloat());
```
We need to test null handling.

benwtrent commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/70_text_similarity_rank_retriever.yml:
```diff
@@ -0,0 +1,90 @@
+setup:
+  - requires:
+      cluster_features: "gte_v8.15.0"
```
```suggestion
      cluster_features: "text_similarity_reranker_retriever_supported"
```
You should be able to use your node feature as declared.

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilderTests.java:
```diff
@@ -0,0 +1,83 @@
+            randomAlphaOfLength(20),
+            randomAlphaOfLength(50),
+            randomIntBetween(1, 10000),
+            randomFloatBetween(-1.0f, 1.0f, true)
```
```suggestion
            randomBoolean() ? null : randomFloatBetween(-1.0f, 1.0f, true)
```

benwtrent replied:
we need to test all null things. If there are other nullable/optional features, please add cases for them.

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankBuilderTests.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.io.IOException;
+
+public class TextSimilarityRankBuilderTests extends AbstractXContentSerializingTestCase<TextSimilarityRankBuilder> {
```
We need a test where `null` rank_window size is parsed and we confirm its the default and that serialization all works.

demjened replied:
Added test case.

benwtrent commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilderTests.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TextSimilarityRankRetrieverBuilderTests extends AbstractXContentTestCase<TextSimilarityRankRetrieverBuilder> {
```
We need a test where `null` rank_window size is parsed and we confirm its the default and that serialization all works.

demjened replied:
Added test case.

benwtrent reviewed (approved):
Thank you for adding all the tests. Overall, this looks good. I added some more comments, we need to test `null` things and make sure that they are handled appropriately.

demjened committed:
Consolidate tests

demjened committed:
Add random nulls and parsing tests with defaults

demjened committed:
Update feature req in yaml test

demjened committed:
Rename preprocess method

demjened committed:
Spotless
